### PR TITLE
Expose OIDC as standalone policy

### PR DESCRIPTION
**Pull request by mikz**
We would like to use OIDC without 3scale, so we need a standalone OIDC policy for authentication.

Requirement for https://github.com/3scale/ostia/issues/14

/cc @3scale/ostia

### CODE REVIEW

File: gateway/src/resty/oidc/discovery.lua [hunk 1]
```diff
@@ -15,6 +15,9 @@ local _M = { }
 
 local mt = { __index = _M }
 
+-- to be overridden in tests
```
Comment (davidor) [inline gateway/src/resty/oidc/discovery.lua]: Why is this needed? Can't we just use the `http_backend` that `.new()` receives?

File: gateway/src/apicast/oauth/oidc.lua [hunk 2]
```diff
@@ -51,7 +51,7 @@ function _M.new(service)
     jwt_claims = {
       -- 1. The JWT MUST contain an "iss" (issuer) claim that contains a
       -- unique identifier for the entity that issued the JWT.
-      iss = jwt_validators.equals_any_of({ issuer }),
+      iss = issuer and jwt_validators.equals_any_of({ issuer }) or jwt_validators.required(),
```
Comment (davidor) [inline gateway/src/apicast/oauth/oidc.lua]: Why is `or jwt_validators.required()` needed?

File: gateway/src/apicast/oauth/oidc.lua [hunk 3]
```diff
@@ -92,64 +95,99 @@ end
...
 
-  local jwt_obj = cache:get(cache_key)
+  jwt.token = token
 
-  if jwt_obj then
+  return setmetatable(jwt, jwt_mt)
+end
+
+function _M:parse(jwt, cache_key)
+  local cached = self.cache:get(cache_key or jwt)
```
Comment (davidor) [inline gateway/src/apicast/oauth/oidc.lua]: Why `cache_key or jwt` ? Shouldn't it be just `cache_key` ?

File: gateway/src/apicast/policy/oidc_authentication/oidc_authentication.lua [hunk 4]
```diff
@@ -0,0 +1,91 @@
+-- This is a oidc_authentication description.
```
Comment (davidor) [inline gateway/src/apicast/policy/oidc_authentication/oidc_authentication.lua]: Interesting :eyes:

File: gateway/src/apicast/policy/oidc_authentication/oidc_authentication.lua [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+local OIDC = require('apicast.oauth.oidc')
+local oidc_discovery = require('resty.oidc.discovery')
+local http_authorization = require('resty.http_authorization')
+local resty_url = require('resty.url')
+local policy = require('apicast.policy')
+local _M = policy.new('oidc_authentication')
+
+local tostring = tostring
+
+_M.cache_size = 100
```
[the remark this answers is not in the document]
Comment (mikz) [reply gateway/src/apicast/policy/oidc_authentication/oidc_authentication.lua]: Other modules to this too. I think it is reasonable. If you want to you could write a policy that in the init phase bumps cache sizes of other policies.

File: gateway/src/apicast/oauth/oidc.lua (hunk 3 above)
Comment (mikz) [reply gateway/src/apicast/oauth/oidc.lua]: Because `cache_key` can be `nil`. So the token itself is its cache key.

File: gateway/src/apicast/oauth/oidc.lua (hunk 2 above)
Comment (mikz) [reply gateway/src/apicast/oauth/oidc.lua]: Basically, I want to validate the issuer, if there is some, or just require the field.

File: gateway/src/resty/oidc/discovery.lua (hunk 1 above)
Comment (mikz) [reply gateway/src/resty/oidc/discovery.lua]: The issue is that when you use this for example in the policy initializer, then the policy should accept this too. Pushing testing client through all levels of the stack seems wrong.

File: gateway/src/apicast/policy/oidc_authentication/oidc_authentication.lua [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+end
+
+local new = _M.new
+--- Initialize a oidc_authentication
+-- @tparam[opt] table config Policy configuration.
+function _M.new(config)
+  local self = new(config)
+
+  self.issuer_endpoint = valid_issuer_endpoint(config and config.issuer_endpoint)
+  self.discovery = oidc_discovery.new(self.http_backend)
```
Comment (davidor) [inline gateway/src/apicast/policy/oidc_authentication/oidc_authentication.lua]: Isn't `self.http_backend` always `nil` at this point?

Comment (mikz) [reply gateway/src/apicast/policy/oidc_authentication/oidc_authentication.lua]: Yes, unless you set it on the module level. `self` already has the metatable pointing to _M.

File: spec/policy/oidc_authentication/oidc_authentication_spec.lua [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+    iss = 'http://example.com/issuer',
+    sub = 'some',
+    aud = 'one',
+    exp = ngx.now() + 3600,
+  },
+}, { __tostring = function(jwt) return JWT:sign(rsa.private, jwt) end })
+
+describe('oidc_authentication policy', function()
+  describe('.new', function()
+    it('works without configuration', function()
```
Comment (davidor) [inline spec/policy/oidc_authentication/oidc_authentication_spec.lua]: For these 2 cases with empty or nil config, it might be useful to check that `rewrite` and `access` do not raise unexpected errors.

File: spec/policy/oidc_authentication/oidc_authentication_spec.lua [hunk 8]
```diff
@@ -0,0 +1,185 @@
...
+
+      stub(oidc_discovery, 'http_backend', test_backend)
+
+      test_backend
+              .expect{ url = issuer_endpoint .. "/.well-known/openid-configuration" }
+              .respond_with{ status = 200, headers = { content_type = 'application/json;charset=UTF-8' },
+                             body = [[
+            {
+              "issuer": "https://example.com/issuer",
+              "jwks_uri": "https://example.com/jwks",
```
Comment (davidor) [inline spec/policy/oidc_authentication/oidc_authentication_spec.lua]: I think that extracting this one to a local var used in both mocks would nicely show how they're related.

File: spec/policy/oidc_authentication/oidc_authentication_spec.lua [hunk 9]
```diff
@@ -0,0 +1,185 @@
...
+      local context = { }
+
+      ngx.var.http_authorization = 'Bearer ' .. tostring(access_token)
+
+      policy.oidc.alg_whitelist = { RS256 = true }
+      policy.oidc.keys = { [access_token.header.kid] = { pem = rsa.pub } }
+
+      policy:rewrite(context)
+
+      assert.contains({ payload = { iss = 'http://example.com/issuer' } }, context.jwt)
```
Comment (davidor) [inline spec/policy/oidc_authentication/oidc_authentication_spec.lua]: Shouldn't this return the other claims? I don't see why it only returns the `iss`.

Comment (mikz) [reply spec/policy/oidc_authentication/oidc_authentication_spec.lua]: It does. But the `assert.contains` verifies just subset. I could extend it to all the claims.

File: gateway/src/apicast/policy/oidc_authentication/oidc_authentication.lua (hunk 6 above)
Comment (davidor) [reply gateway/src/apicast/policy/oidc_authentication/oidc_authentication.lua]: :+1:

File: gateway/src/resty/oidc/discovery.lua (hunk 1 above)
Comment (davidor) [reply gateway/src/resty/oidc/discovery.lua]: If this is only for testing, another option would be to modify `self.backend` in the tests. Not sure if better.

I get your point of having to pass this through multiple objects.

:+1:

Comment (mikz) [reply gateway/src/resty/oidc/discovery.lua]: Actually, this got me thinking. I think we should offer this in http_ng, so we can stub it globally in the whole test suite.

File: spec/policy/oidc_authentication/oidc_authentication_spec.lua (hunk 9 above)
Comment (davidor) [reply spec/policy/oidc_authentication/oidc_authentication_spec.lua]: I'd do that. It's a bit confusing in my opinion right now.
How about `assert.same(access_token.payload, context.jwt)` ?

File: gateway/src/apicast/oauth/oidc.lua (hunk 2 above)
Comment (davidor) [reply gateway/src/apicast/oauth/oidc.lua]: Hmm so is it OK if `iss` is there but not included in `issuer` ? :thinking:

Comment (mikz) [reply gateway/src/apicast/oauth/oidc.lua]: Well, that's the thing. If there is no issuer in the configuration, what to do?

I decided to just require any value for `iss`.

File: spec/policy/oidc_authentication/oidc_authentication_spec.lua [hunk 10]
```diff
@@ -0,0 +1,185 @@
...
+      _M.new{ issuer_endpoint = issuer_endpoint }
+    end)
+  end)
+
+  describe(':rewrite', function()
+    before_each(function()
+      ngx.var = {}
+    end)
+
+    it('parses access token', function()
```
Comment (davidor) [inline spec/policy/oidc_authentication/oidc_authentication_spec.lua]: I'd also mention that it stores the parsed token in the context.
I think this is important because other policies might want to check if there's already a parsed jwt in there.

**Comment by davidor** [comment]
Shouldn't this policy have an `apicast-policy.json` ?

**Comment by mikz** [comment]
That would show it in the UI. I'm not sure we are ready for that. What would happen when a customer would add this policy to existing OIDC service? IMO we should think a bit more how that should work before exposing it.
